### ISSUE TITLE
Segmentation fault on reconnect_with_callback function (Ubuntu 18.04/Mojave)

### ISSUE THREAD

**Issue by AdrienJonquet** [eclipse/paho.mqtt.rust#33]
A segmentation faults occurs when accessing the connection options (`opts`) in the `reconnect_with_callback` function when I turn off wifi or unplug ethernet.

```
    pub fn reconnect_with_callbacks<FS,FF>(&self,
                                           success_cb: FS,
                                           failure_cb: FF) -> Arc<Token>
        where FS: FnMut(&AsyncClient,u16) + 'static,
              FF: FnMut(&AsyncClient,u16,i32) + 'static
    {
        let connopts = {
            let lkopts = self.inner.opts.lock().unwrap();
            (*lkopts).clone()
        };
        self.connect_with_callbacks(connopts, success_cb, failure_cb)
    }
```
`self.inner.opts.lock().unwrap()` => opts is NULL

I reproduce this issue on both Ubuntu 18.04 and macos Mojave using regular asynchronous examples. The `connect_with_callbacks` works very well though when re-executing the program without internet.

**Comment by fpagliughi** [comment]
Thanks. I'll have a look.

**Comment by ckalas** [comment]
any update on this? I am facing this same problem now for the last few days and can't resolve it

**Comment by fpagliughi** [comment]
I was able to reproduce this bug. It appears that I'm sending a reference to a temporary _AsyncClient_ object to the _on_connect_failure()_ callback. I'll see if I can get a quick fix in there, but in the longer run, I should probably change the API to use Arc client as I just mused about this morning in Issue #31.

**Comment by fpagliughi** [comment]
A fix is in the _develop_ branch, and should go out with the next update.

**Comment by joeferner** [comment]
I can reproduce it as well.

1. Start mqtt broker (mosquitto)
2. use `AsyncClient.connect_with_callbacks`
3. In the on connect callback `client.subscribe`
4. everything appears to work
5. stop mqtt broker
6. start mqtt broker
7. fault occurs when calling subscribe in the on connect success callback (https://github.com/joeferner/unisonht-raspberry-pi-ir-hat/blob/949450a980ff3d545b106f71aa23db7fc1280c1d/src/main.rs#L225)

vscode has it breaking on this line... https://github.com/eclipse/paho.mqtt.rust/blob/master/src/async_client.rs#L671

Looks like `inner` is disposed.

**Comment by mfanara** [comment]
Regarding the comment by [joeferner](https://github.com/joeferner)  on Nov 6, 2021. I see the exact same thing when using unmodified examples/legacy_async_subscribe.

**Comment by fpagliughi** [comment]
It appears that there was an issue with earlier version of the Paho C library which got fixed, then a new one popped up, causing a similar problem. The next release will concentrate on getting this once and for all.

**Comment by fpagliughi** [comment]
It looks like this is _finally_ fixed with the upcoming v0.12.3 release. At least I can't recreate it in testing.
If you still see it with that version, please feel free to reopen.
